test/demos/reversediffzero.jl

```diff
@@ -0,0 +1,132 @@
+"The simplest viable reverse mode a AD, only supports `Float64`"
+module ReverseDiffZero
+using ChainRulesCore
+using Test
+
+struct Tracked <: Real
+    propagate::Function
+    primal::Float64
+    tape::Vector{Any}  # a reference to a shared tape
+    grad::Base.RefValue{Float64} # current accumulated gradient
+end
+
+"An intermediate value, a Branch in Nabla terms."
+function Tracked(propagate, primal, tape)
+    v = Tracked(propagate, primal, tape, Ref(zero(primal)))
+    push!(tape, v)
+    return v
+end
+
+"An input, a Leaf in Nabla terms. No inputs of its on to propagate to."
+function Tracked(primal, tape)
+    # don't actually need to put these on the tape, since they don't need to
+    # propagate anything
+    return Tracked(_->nothing, primal, tape, Ref(zero(primal)))
+end
+
+primal(d::Tracked) = d.primal
+primal(d) = d
+
+grad(d::Tracked) = d.grad[]
+grad(d) = nothing
+
+tape(d::Tracked) = d.tape
+tape(d) = nothing
+
+# we have many inputs grab the tape from the first one that is tracked
+get_tape(ds) = something(tape.(ds)...)
+
+# propagate the currently stored gradient back to my inputs.
+propagate!(d) = nothing
+propagate!(d::Tracked) = d.propagate(d.grad[])
+
+# Accumulate gradient, if the value is being tracked.
+accum!(d, x̄) = nothing
+accum!(d::Tracked, x̄) = d.grad[] += x̄
```

Comment:
Just for consistency

```suggestion
accum!(d::Tracked, x̄) = d.grad[] += x̄
accum!(d, x̄) = nothing
```